echopype/process/process_base.py

```diff
@@ -260,33 +266,39 @@ class ProcessEK(ProcessBase):
     def __init__(self, model=None):
         super().__init__(model)
 
-    def calc_cal_correction(self, ed, param='sa_correction'):
-        """ Calculate the sa correction from the sa correction table and the pulse length.
-        For EK80 BB mode, there is no sa correction table so the sa correction is saved as none
+    @staticmethod
+    def get_power_cal_params(ed, param='sa_correction'):
+        """Get calibration parameters for power/angle data.
         """
+        # TODO: need to test with EK80 power/angle data
+        #  currently this has only been tested with EK60 data
         ds_vend = ed.get_vend_from_raw()
 
         if param not in ds_vend:
             return None
 
-        if param == 'sa_correction':
-            correction_table = ds_vend.sa_correction
-        elif param == 'gain_correction':
-            correction_table = ds_vend.gain_correction
-        else:
+        if param not in ['sa_correction', 'gain_correction']:
             raise ValueError(f"Unknown parameter {param}")
 
-        pulse_length_table = ds_vend.pulse_length
-        unique_pulse_length = np.unique(ed.raw.transmit_duration_nominal, axis=1).squeeze()
-
-        if unique_pulse_length.size != ds_vend.frequency.size:
-            raise ValueError("Pulse length changes over time")
-        idx = [np.argwhere(np.isclose(unique_pulse_length[i], pulse_length_table[i])).squeeze()
-               for i in range(pulse_length_table.shape[0])]
-
-        correction = np.array([ch[x] for ch, x in zip(correction_table, np.array(idx))])
-
-        return xr.DataArray(correction, dims='frequency').assign_coords(frequency=correction_table.frequency)
+        # Drop NaN ping_time for transmit_duration_nominal
+        if np.any(np.isnan(ed.raw['transmit_duration_nominal'])):
+            # TODO: resolve performance warning:
+            #  /Users/wu-jung/miniconda3/envs/echopype_jan2021/lib/python3.8/site-packages/xarray/core/indexing.py:1369:
+            #  PerformanceWarning: Slicing is producing a large chunk. To accept the large
+            #  chunk and silence this warning, set the option
+            #      >>> with dask.config.set(**{'array.slicing.split_large_chunks': False}):
+            #      ...     array[indexer]
+            #  To avoid creating the large chunks, set the option
+            #      >>> with dask.config.set(**{'array.slicing.split_large_chunks': True}):
+            #      ...     array[indexer]
+            #    return self.array[key]
+            ed.raw = ed.raw.dropna(dim='ping_time', how='any', subset=['transmit_duration_nominal'])
```

Comment:
This might fit better in `Echodata `since it is unexpected for the raw to change during `init_cal_params`

Reply:
Good point. Seem that we can have an echopype.qc to clean up issues like this and the ping_time going backward issue.